### PR TITLE
libct/cg/sd/v1: Fix unnecessary freeze/thaw

### PR DESCRIPTION
**Pull request by odinuge**
My testing shows that https://github.com/opencontainers/runc/pull/3082 (commit https://github.com/opencontainers/runc/pull/3082/commits/f2db87986cab5a7c98b8d9b875bb54e0e7a44219) doesn't work as expected. This causes big issues in k8s.

PTAL @kolyshkin :smile: 

Finally back again after having some time off, so haven't had time to look at it before now.

### CODE REVIEW

File: libcontainer/cgroups/systemd/common.go [hunk 1]
```diff
@@ -518,3 +522,19 @@ func addCpuset(cm *dbusConnManager, props *[]systemdDbus.Property, cpus, mems st
 	}
 	return nil
 }
+
+func getUnitType(unitName string) string {
```
Comment (odinuge) [inline libcontainer/cgroups/systemd/common.go:526]: Placeholder. Can probably do something more smooth. This is just so we can start discussing the best way to solve it.

File: libcontainer/cgroups/systemd/v1.go [hunk 2]
```diff
@@ -355,13 +356,20 @@ func (m *legacyManager) freezeBeforeSet(unitName string, r *configs.Resources) (
...
+		devPolicy, e := getUnitTypeProperty(m.dbus, unitName, getUnitType(unitName), "DevicePolicy")
 		if e == nil && devPolicy.Value == dbus.MakeVariant("auto") {
-			devAllow, e := getUnitProperty(m.dbus, unitName, "DeviceAllow")
-			if e == nil && devAllow.Value == dbus.MakeVariant([]deviceAllowEntry{}) {
-				needsFreeze = false
-				needsThaw = false
-				return
+			devAllow, e := getUnitTypeProperty(m.dbus, unitName, getUnitType(unitName), "DeviceAllow")
+			if e == nil {
+				rv := reflect.ValueOf(devAllow.Value.Value())
```
Comment (kolyshkin) [inline libcontainer/cgroups/systemd/v1.go:363]: Do you mean that `devAllow.Value == dbus.MakeVariant([]deviceAllowEntry{})` did not work? Or that your version is more readable? (or both?)

File: libcontainer/cgroups/systemd/common.go (hunk 1 above)
Comment (kolyshkin) [reply libcontainer/cgroups/systemd/common.go:526]: We can probably just use `strings.ToTitle()` here.

File: libcontainer/cgroups/systemd/v1.go (hunk 2 above)
Comment (odinuge) [reply libcontainer/cgroups/systemd/v1.go:363]: > Do you mean that devAllow.Value == dbus.MakeVariant([]deviceAllowEntry{}) did not work? Or that your version is more readable? (or both?)

Well, it doesn't seen to work for me. Does that for you, in that case there is something wrong with my setup? :thinking:

File: libcontainer/cgroups/systemd/common.go (hunk 1 above)
Comment (odinuge) [reply libcontainer/cgroups/systemd/common.go:526]: Yeah, was thinking about that. Or should we have an allowlist like this, to ensure we don't forward "bad" strings? Doesn't matter that much to me

Comment (odinuge) [reply libcontainer/cgroups/systemd/common.go:526]:
```go
validStuff := map[string]bool {
     "Slice": true
     "Scope": true
 }
  
 if validStuff[strings.ToTitle(type)] {
     return type
 }
 return err
```

Can probably do something like this to validate it.

**Comment by kolyshkin** [comment]
I guess what we need first is a unit test that would create a unit (of various types) with a particular property, get the property, and check it's as expected.

Last commit of #3082 is an integration test (`TestPodSkipDevicesUpdate`) that checks that the freeze is not done. I look at all this and I still can't understand why that test does not catch the issue this PR is fixing...

**Comment by odinuge** [comment]
>I guess what we need first is a unit test that would create a unit (of various types) with a particular property, get the property, and check it's as expected.

Yeah, some tests would be ok I guess. It is fairly hard to test tho. But keep reading `freezer.state` on the slice while constantly updating, and looking for only THAWED would work. I can look into doing it in go, but the bats stuff is a bit complicated tho., so that would require some boiler plate.

> Last commit of #3082 is an integration test (TestPodSkipDevicesUpdate) that checks that the freeze is not done. I look at all this and I still can't understand why that test does not catch the issue this PR is fixing...

Well, it only verifies that there is no errors with the devices. It does however not check if the slice is frozen or not.


To test this, one can simply verify like this - keep the loop running while running tests with `sudo go test -run=TestPodSkipDevicesUpdate -v -test.count=5`:

```
# in this branch
$ while true ; do grep FROZEN -R /sys/fs/cgroup/freezer/system.slice/; done
/sys/fs/cgroup/freezer/system.slice/system-runc_test_podTestPodSkipDevicesUpdate.slice/test-PodSkipDevicesUpdate.scope/freezer.state:FROZEN
/sys/fs/cgroup/freezer/system.slice/system-runc_test_podTestPodSkipDevicesUpdate.slice/test-PodSkipDevicesUpdate.scope/freezer.state:FROZEN
# in master 
$ while true ; do grep FROZEN -R /sys/fs/cgroup/freezer/system.slice/; done             
/sys/fs/cgroup/freezer/system.slice/system-runc_test_podTestPodSkipDevicesUpdate.slice/test-PodSkipDevicesUpdate.scope/freezer.state:FROZEN                     
/sys/fs/cgroup/freezer/system.slice/system-runc_test_podTestPodSkipDevicesUpdate.slice/freezer.state:FROZEN                                                     
/sys/fs/cgroup/freezer/system.slice/system-runc_test_podTestPodSkipDevicesUpdate.slice/test-PodSkipDevicesUpdate.scope/freezer.state:FROZEN                     
/sys/fs/cgroup/freezer/system.slice/system-runc_test_podTestPodSkipDevicesUpdate.slice/freezer.state:FROZEN                                                     
/sys/fs/cgroup/freezer/system.slice/system-runc_test_podTestPodSkipDevicesUpdate.slice/test-PodSkipDevicesUpdate.scope/freezer.state:FROZEN                     
/sys/fs/cgroup/freezer/system.slice/system-runc_test_podTestPodSkipDevicesUpdate.slice/freezer.state:FROZEN                                                     
/sys/fs/cgroup/freezer/system.slice/system-runc_test_podTestPodSkipDevicesUpdate.slice/test-PodSkipDevicesUpdate.scope/freezer.state:FROZEN                     
/sys/fs/cgroup/freezer/system.slice/system-runc_test_podTestPodSkipDevicesUpdate.slice/freezer.state:FROZEN                                                     
/sys/fs/cgroup/freezer/system.slice/system-runc_test_podTestPodSkipDevicesUpdate.slice/test-PodSkipDevicesUpdate.scope/freezer.state:FROZEN                     
/sys/fs/cgroup/freezer/system.slice/system-runc_test_podTestPodSkipDevicesUpdate.slice/freezer.state:FROZEN                                                     
/sys/fs/cgroup/freezer/system.slice/system-runc_test_podTestPodSkipDevicesUpdate.slice/test-PodSkipDevicesUpdate.scope/freezer.state:FROZEN                     
/sys/fs/cgroup/freezer/system.slice/system-runc_test_podTestPodSkipDevicesUpdate.slice/freezer.state:FROZEN                                                     
/sys/fs/cgroup/freezer/system.slice/system-runc_test_podTestPodSkipDevicesUpdate.slice/test-PodSkipDevicesUpdate.scope/freezer.state:FROZEN                     
/sys/fs/cgroup/freezer/system.slice/system-runc_test_podTestPodSkipDevicesUpdate.slice/freezer.state:FROZEN                                                     
/sys/fs/cgroup/freezer/system.slice/system-runc_test_podTestPodSkipDevicesUpdate.slice/test-PodSkipDevicesUpdate.scope/freezer.state:FROZEN                     
/sys/fs/cgroup/freezer/system.slice/system-runc_test_podTestPodSkipDevicesUpdate.slice/freezer.state:FROZEN                                                     
/sys/fs/cgroup/freezer/system.slice/system-runc_test_podTestPodSkipDevicesUpdate.slice/freezer.state:FROZEN                        
```

As seen here, only the scope/container is frozen with this branch, while in master both the pod/slice and the container/scope is frozen. Does that make sense to you @kolyshkin ?

**Comment by odinuge** [comment]
Hmm. Is something wrong with CI, or?

**Comment by kolyshkin** [comment]
> Hmm. Is something wrong with CI, or?

Yup; see https://githubstatus.com/.

Seems over now, do a force push to trigger CI.

**Comment by kolyshkin** [comment]
or we can close/open to re-kick CI (usually it works)

**Comment by AkihiroSuda** [comment]
Does this fix https://github.com/kubernetes/kubernetes/issues/104280 ?

**Comment by odinuge** [comment]
> @AkihiroSuda: Does this fix kubernetes/kubernetes#104280 ?

Yes.

But as I have said a few times already, for k8s a flag for disabling this would be better and simpler. I guess we (in k8s, not here in runc) have to discuss if we should avoid using `runc/libcontainer` for setting cgroup things, since things like this will increase the overhead waaay more than necessary.

File: libcontainer/cgroups/systemd/common.go (hunk 1 above)
Comment (kolyshkin) [reply libcontainer/cgroups/systemd/common.go:526]: We can either have `.slice` or `.scope` (see `getUnitName()`).

I suggest we add `getUnitType` which is very similar to `getUnitName`.

File: libcontainer/cgroups/systemd/v1.go (hunk 2 above)
Comment (kolyshkin) [reply libcontainer/cgroups/systemd/v1.go:366]: I think I improved the readability a notch in https://github.com/opencontainers/runc/pull/3148:

```go
                        devAllow, e := getUnitTypeProperty(m.dbus, unitName, unitType, "DeviceAllow")
                          if e == nil {
                                  if rv := reflect.ValueOf(devAllow.Value.Value()); rv.Kind() == reflect.Slice && rv.Len() == 0 {
                                          needsFreeze = false
                                          needsThaw = false
                                          return
                                  }
                          }
```

**Comment by kolyshkin** [comment]
@odinuge I was playing with the test case and ended up doing https://github.com/opencontainers/runc/pull/3148.

The other thing I did is slightly simplified the deviceAllow check (https://github.com/opencontainers/runc/pull/3143#discussion_r687222469)

We can either use #3148, or cherry-pick the test case from it to here.

File: libcontainer/cgroups/systemd/v1.go (hunk 2 above)
Comment (odinuge) [reply libcontainer/cgroups/systemd/v1.go:366]: Guess I can do that instead if the others agree, even tho. the line is a bit too long. I'll update the PR.

File: libcontainer/cgroups/systemd/common.go [hunk 3]
```diff
@@ -518,3 +518,10 @@ func addCpuset(cm *dbusConnManager, props *[]systemdDbus.Property, cpus, mems st
...
```
Comment (kolyshkin) [inline libcontainer/cgroups/systemd/common.go:522]: Can you please add something like
```go
	// This code should be in sync with getUnitName.
```

**Comment by odinuge** [comment]
> We can either use #3148, or cherry-pick the test case from it to here.

I can cherry-pick the test commit here if they should be in the same PR. I prefer to keep this PR, otherwise I cannot update and change the PR (you created based of this), making it a bit cumbersome to work on it. Having a separate PR with the test works as well.

File: libcontainer/cgroups/systemd/common.go (hunk 3 above)
Comment (odinuge) [reply libcontainer/cgroups/systemd/common.go:522]: Ack 👍

**Comment by kolyshkin** [comment]
> I can cherry-pick the test commit here if they should be in the same PR

Please do (otherwise the CI is semi-useless).

> making it a bit cumbersome to work on it

Something like `git-rebase -i origin/master` is a life saver when working on PRs with multiple commits. In most complicated cases I do `git stash; git format-patch -o <dir> origin/master; git reset --hard origin/master`, when `git am` those commits one by one, making changes and doing  `git commit --amend` when necessary.

File: libcontainer/cgroups/systemd/v1.go (hunk 2 above)
Comment (kolyshkin) [reply libcontainer/cgroups/systemd/v1.go:366]: What about this:
```go
                        if e == nil { 
                                rv := reflect.ValueOf(devAllow.Value.Value())
                                if rv.Kind() == reflect.Slice && rv.Len() == 0 { 
```

Comment (kolyshkin) [reply libcontainer/cgroups/systemd/v1.go:366]: To be it reads slightly better than the switch/case (although I agree that either version we came with so far is not super readable).

**Comment by kolyshkin** [comment]
Once ready and merged, we'll have a backport to 1.0 and will try 1.0.2 released ASAP.

**Comment by kolyshkin** [comment]
> things like this will increase the overhead waaay more than necessary

@odinuge your concern about this code path being slow is valid, it is also quite complicated. I was looking into how to make it fast and simple -- PTAL https://github.com/opencontainers/runc/pull/3151.

Surely a simple "SkipFreeze" flag will be faster, but this seems lightweight enough to not cause performance concerns.

**Comment by odinuge** [comment]
> @odinuge your concern about this code path being slow is valid, it is also quite complicated. I was looking into how to make it fast and simple -- PTAL #3151.
>
> Surely a simple "SkipFreeze" flag will be faster, but this seems lightweight enough to not cause performance concerns.

Yeah, #3151 would probably work, but I don't like going around systemd like that approach does. Systemd should be the source of the truth in this case, as it is now.

(agian, I am mostly a fan of a flag for k8s, otherwise we have to look for alternatives. k8s do these calls very often, and with an increasing amount of pods, this will increase further. Same for shortlived containers. A flag as `SkipDevices` was previously is definitely the best approach for k8s; whether we have to use runc or implement it ourselves.)

> Something like git-rebase -i origin/master is a life saver when working on PRs with multiple commits. In most complicated cases I do git stash; git format-patch -o <dir> origin/master; git reset --hard origin/master, when git am those commits one by one, making changes and doing git commit --amend when necessary.

`git` is always the easy part. GitHub is the hard one, since I don't have push or edit access on your PR, only mine.

**Comment by harche** [comment]
/hold 

I am running node e2e + systemd + crio tests locally using the runc binary built from this PR. The first run was successful, but I will have more runs to verify that this PR indeed fixes https://github.com/kubernetes/kubernetes/issues/104280#issuecomment-897093892

**Comment by odinuge** [comment]
> I am running node e2e + systemd + crio tests locally using the runc binary built from this PR. The first run was successful, but I will have more runs to verify that this PR indeed fixes kubernetes/kubernetes#104280 (comment)

It is the vendoring of runc into k8s that is important for fixing kubernetes/kubernetes#104280, not only the runc binary used by the CRI runtime.

**Comment by harche** [comment]
> > I am running node e2e + systemd + crio tests locally using the runc binary built from this PR. The first run was successful, but I will have more runs to verify that this PR indeed fixes [kubernetes/kubernetes#104280](https://github.com/kubernetes/kubernetes/issues/104280) (comment)
> 
> It is the vendoring of runc into k8s that is important for fixing [kubernetes/kubernetes#104280](https://github.com/kubernetes/kubernetes/issues/104280), not only the runc binary used by the CRI runtime.

True, but the failures I was seeing in CI were specifically coming from runc binary indicating that CRIO was able to start the container but it ended up in paused state. 

/hold cancel

**Comment by odinuge** [comment]
> True, but the failures I was seeing in CI were specifically coming from runc binary indicating that CRIO was able to start the container but it ended up in paused state.

Yes, but that is due to kubelet (via this code) freezing `kubepods.slice`, `burstable.slice`, and `pod-xyz.slice`. Most of the issues in regards to the binary should already be fixed in runc `v1.0.1`.

**Comment by harche** [comment]
> > True, but the failures I was seeing in CI were specifically coming from runc binary indicating that CRIO was able to start the container but it ended up in paused state.
> 
> Yes, but that is due to kubelet (via this code) freezing `kubepods.slice`, `burstable.slice`, and `pod-xyz.slice`. Most of the issues in regards to the binary should already be fixed in runc `v1.0.1`.

Oh ok, that means the success I am seeing in node e2e (2 out of 2 runs so far) is due to me updating runc from rc95 to your latest PR (which has v1.0.1).

File: libcontainer/cgroups/systemd/v1.go [hunk 4]
```diff
@@ -353,15 +354,20 @@ func (m *legacyManager) freezeBeforeSet(unitName string, r *configs.Resources) (
 		// a non-existent unit returns default properties,
 		// and settings in (2) are the defaults.
```
Comment (kolyshkin) [inline libcontainer/cgroups/systemd/v1.go:355]: The tests in this PR do not check whether this is true. I have added those checks last night, plus another test case -- @odinuge can you please re-cherry-pick the last commit from https://github.com/kolyshkin/runc/tree/fix-freeze-before-set

**Comment by kolyshkin** [comment]
> Yeah, #3151 would probably work, but I don't like going around systemd like that approach does. Systemd should be the source of the truth in this case, as it is now.

So, we have to admit that this is a kludge already. We need a way to figure out whether systemd is going do a disruptive device rules update. Any way to figure that out is fine as long as it works. Checking unit properties is one way, checking devices.list is another. Sure, #3151 has an assumption that devices.list in the kernel and systemd Device* properties are in sync. I think such assumption is valid (if not, this would mean systemd is not doing its job, ignoring device properties updates or some such).

> (agian, I am mostly a fan of a flag for k8s, otherwise we have to look for alternatives. k8s do these calls very often, and with an increasing amount of pods, this will increase further. Same for shortlived containers. A flag as `SkipDevices` was previously is definitely the best approach for k8s; whether we have to use runc or implement it ourselves.)

Yes, this is also a way to figure out whether a freeze is needed or not, and it's fine as long as it works. The only thing I don't like about it is it requires an additional knob to set. Generally speaking, if there is a way to achieve a goal without adding an explicit knob, and it works reliably and fast, I think such way is better. Currently I see #3151 as such a way.

**Comment by odinuge** [comment]
> So, we have to admit that this is a kludge already. We need a way to figure out whether systemd is going do a disruptive device rules update. Any way to figure that out is fine as long as it works. Checking unit properties is one way, checking devices.list is another. Sure, #3151 has an assumption that devices.list in the kernel and systemd Device* properties are in sync. I think such assumption is valid (if not, this would mean systemd is not doing its job, ignoring device properties updates or some such).

Yeah, that solution does also work, and yeah, this is all pretty kludgy, I agree about that :sweat_smile: k8s and the runc binary also work differently, with different purposes, making it even more kludgy. But yeah, then we instead can just look at the provided resources, and if the devices allow access to all, then drop the freeze. If we don't rely on systemd, and don't want a separate knob, I think checking devices would be better and simpler to understand, together with better performance.

Eg something like:

```go
oldDevices :=  []specs.LinuxDeviceCgroup{
  {
    Allow:  true,
    Access: "rwm",
  },
}

newDevices :=  []specs.LinuxDeviceCgroup{
  {
    Allow:  true,
    Access: "rwm",
  },
}


if allowsAll(oldDevices) && allowsAll(newDevices) {
  // Skip freezing
}
```

(although we might need to poke systemd or cgropufs in `Apply` to get state in the beginning, in slice/scope is already alive)

Any thoughts about that instead? Wouldn't that work for `runc update` as well, as long as people only manage the slice/scope via runc and not set devices directly in systemd? Or do you think that is too fragile?

> Yes, this is also a way to figure out whether a freeze is needed or not, and it's fine as long as it works. The only thing I don't like about it is it requires an additional knob to set. Generally speaking, if there is a way to achieve a goal without adding an explicit knob, and it works reliably and fast, I think such way is better. Currently I see #3151 as such a way.

The `SkipDevices` is _already_ a knob, and it worked properly for k8s until it was adapted for `runc update`. I don't really see an issue with it being a knob in the same way as `SkipDevices` is, but I guess that is another discussion. :sweat_smile:

File: libcontainer/cgroups/systemd/v1.go (hunk 4 above)
Comment (odinuge) [reply libcontainer/cgroups/systemd/v1.go:355]: Ack :smile: Test looks good!

**Comment by odinuge** [comment]
We can also change this to a single dbus call by using `GetUnitTypePropertiesContext` on the slice/scope, since both `DeviceAllow` and `DevicePolicy` are on the same interface.

**Comment by kolyshkin** [comment]
> We can also change this to a single dbus call by using GetUnitTypePropertiesContext on the slice/scope, since both DeviceAllow and DevicePolicy are on the same interface.

It's an interesting idea, which I also considered but gave up on that since this returns all (tens of, I guess) properties in a `map[string]interface{}`, of which we only need two. This results in more garbage to collect by the GC.

OTOH if the call itself is relatively heavy, the cost of the saved call may outweigh the GC overhead, and this might be worth doing.

I guess the way to check which way is better is to benchmark both implementations, using runtime.GC() inside the benchmarked code to make sure GC overhead is accounted for. IOW something like

```go
func BenchmarkXXX(b *testing.B) {
     // ... any prep code
     runtime.GC()
     b.ResetTimer()
     for i := 0; i < b.N, i++ {
          // ... whatever we test
     }
     runtime.GC()
}
```

**Comment by kolyshkin** [comment]
> ```go
> if allowsAll(oldDevices) && allowsAll(newDevices) {
>   // Skip freezing
> }
> ```
>
> (although we might need to poke systemd or cgropufs in `Apply` to get state in the beginning, in slice/scope is already alive)
> 
> Any thoughts about that instead? Wouldn't that work for `runc update` as well, as long as people only manage the slice/scope via runc and not set devices directly in systemd? Or do you think that is too fragile?

Well, this is conceptually the same that we do here and/or in #3151. In terms of #3151, `allowsAll(oldDevices)` is equivalent to `!hasDeviceRules()`, and `allowsAll(newDevices)` is equivalent to `r.SkipDevices`.

**Comment by kolyshkin** [comment]
> `git` is always the easy part. GitHub is the hard one, since I don't have push or edit access on your PR, only mine.

Ha! This is easy peasy  -- become a runc maintainer, and you'll be able to shove your commits into contributors' PRs 🕺🏻 

On a serious note, though, I think I misunderstood you, thought you were against cherry-picking a commit. 😀

**Comment by kolyshkin** [comment]
OK I ended up writing this benchmark for `freezeBeforeSet()` (based on the test code):

Here are the results (first line is this PR, second is #3151):
```
BenchmarkFreezeBeforeSet-8   	    3757	    286971 ns/op	   16632 B/op	     680 allocs/op
BenchmarkFreezeBeforeSet-8   	  131257	      9116 ns/op	    2344 B/op	       7 allocs/op
```

To have some reference points, I did a few quick benchmarks

1. start/stop a systemd scope:
```
BenchmarkApplyDestroy-8      	     808	   1390820 ns/op	  172188 B/op	    2130 allocs/op
``` 

2. Set() (with no freeze checking or freeze code -- as if SkipFreeze flag is provided)
```
BenchmarkSet-8               	    6412	    183464 ns/op	    7266 B/op	     283 allocs/op
```

3. Set() (with unconditional freeze, no processes in cgroup!):
```
BenchmarkSet-8               	    5127	    231167 ns/op	   10224 B/op	     298 allocs/op
```

From that we can estimate Set() overhead for kubelet case

* with current approach (#3082 (commit f2db879) + #3143): slowdown from 200 to 500 us (150% worse)
* with devices.list approach (#3151): slowdown from 200 to 210 us (5% worse)

**Comment by kolyshkin** [comment]
Ah, here's the benchmarking code: https://gist.github.com/kolyshkin/20e8f43fedde4989033941be8bef9fc6

It needs some tweaks to measure Set(), I was using these placeholders for Set measurements:

```go
func (m *legacyManager) freezeBeforeSetNo(unitName string, r *configs.Resources) (needsFreeze, needsThaw bool, err error) {
        return false, false, nil
}

func (m *legacyManager) freezeBeforeSetYes(unitName string, r *configs.Resources) (needsFreeze, needsThaw bool, err error) {
        return true, true, nil
}
```

**Comment by odinuge** [comment]
> With current approach (#3082 (commit f2db879) + #3143): slowdown from 200 to 500 us (150% worse).
>
> With devices.list approach (#3151): slowdown from 200 to 210 us (5% worse)

Nice. I guess that is more or less predicted. Freezing with a lot of procs, eg. 110 pods on 100 logical Linux CPUs, I assume the overhead will be quite noticeable, and probably quite huge (so people don't infer that unconditional freeze works ok).

> Well, this is conceptually the same that we do here and/or in #3151. In terms of #3151, allowsAll(oldDevices) is equivalent to !hasDeviceRules(), and allowsAll(newDevices) is equivalent to r.SkipDevices.

Well, looking at it that way all the solutions are conceptually the same. But if we only look at the oci-data to see if we can avoid freezing, without any external calls. `allowsAll(oldDevices)` would then just look at the old value, instead of checking with systemd or looking at cgroupfs to verify it.

**Comment by kolyshkin** [comment]
> Well, looking at it that way all the solutions are conceptually the same. But if we only look at the oci-data to see if we can avoid freezing, without any external calls. `allowsAll(oldDevices)` would then just look at the old value, instead of checking with systemd or looking at cgroupfs to verify it.

Problem is, runc cgroup manager do not have persistent storage, and kubelet instantiates a new cgroup manager to call Set():

https://github.com/kubernetes/kubernetes/blob/b6d83f0ba3f155a4f6af9b37bf5511f66327cd5b/pkg/kubelet/cm/cgroup_manager_linux.go#L487

**Comment by odinuge** [comment]
> Problem is, runc cgroup manager do not have persistent storage, and kubelet instantiates a new cgroup manager to call Set():
>
> https://github.com/kubernetes/kubernetes/blob/b6d83f0ba3f155a4f6af9b37bf5511f66327cd5b/pkg/kubelet/cm/cgroup_manager_linux.go#L487

Yeah, but we do supply cgroup config when creating the manager. But yeah, if the only way to avoid this "freeze dance" is for k8s to keep the manager around, I guess we can do that; but that will require some more code.

(not familiar with how the runc binary works in detail btw.)

**Comment by kolyshkin** [comment]
> not familiar with how the runc binary works in detail btw.

It's (usually) a short-lived binary, so we also instantiate a new manager for e.g. `runc update` or `runc destroy`. runc has a state file though (which I plan to split into two parts), so it "knows" the current state (unless someone uses other tools to change it, that is).

**Comment by kolyshkin** [comment]
> Yeah, but we do supply cgroup config when creating the manager.

You can supply anything, meaning this solution would be a skip-freeze flag in disguise.

Oh well, computers are tough. 😩 

https://github.com/opencontainers/runc/pull/3161

**Comment by odinuge** [comment]
> You can supply anything, meaning this solution would be a skip-freeze flag in disguise.

And that is exactly what I want! :smile: 

> Oh well, computers are tough.

Yeee, they are...

> #3161

Ack. +1 for that one

**Comment by kolyshkin** [comment]
Superceded by (already merged) https://github.com/opencontainers/runc/pull/3166 (which includes commits from this PR), thus closing. Thanks @odinuge for figuring this out!

**Pull request closed without merging**